addons/website/static/src/components/fields/fields.js

```diff
@@ -16,26 +16,49 @@ const {Component, useState, onWillStart} = owl;
  */
 class PageUrlField extends Component {
     setup() {
+        this.orm = useService('orm');
+        this.serverUrl = `${window.location.origin}/`;
+        this.pageUrl = this.fieldURL(this.props.value);
+        this.pageMimetypes = {};
+
         this.state = useState({
             redirect_old_url: false,
             url: this.props.value,
             redirect_type: '301',
         });
-        useInputField({getValue: () => this.props.value.url || this.props.value});
 
-        this.serverUrl = window.location.origin;
-        this.pageUrl = this.props.value;
+        onWillStart(async () => {
+            this.pageMimetypes = await this.orm.call('website', 'guess_mimetype', []);
+        });
+        useInputField({getValue: () => this.fieldURL(this.props.value.url || this.props.value)});
     }
 
     get enableRedirect() {
-        return this.state.url !== this.pageUrl;
+        return this.fieldURL(this.state.url, true) !== this.pageUrl;
     }
 
     onChangeRedirectOldUrl(value) {
         this.state.redirect_old_url = value;
         this.updateValues();
     }
 
+    fieldURL(url, slugify) {
+        if (slugify) {
+            url = url.split('/').reduce((path, section, index, sections) => {
+                let slugified = _.str.slugify(section);
+                if (sections.length - 1 === index) {
+                    const i = section.lastIndexOf('.');
+                    const extension = section.slice(i);
+                    if (i > 0 && this.pageMimetypes[extension]) {
+                        slugified = _.str.slugify(section.slice(0, i)) + extension;
+                    }
+                }
+                return slugified ? `${path}/${slugified}` : path;
+            }, '');
+        }
+        return url.replace(/^\//g, '');
+    }
```

Comment:
Could you use some comments and function documentations. Was the "slugification" done in JS like that in 15.0? Trying to remember how this got lost :) cc @rdeodoo Maybe that was done in python but finally doing it in JS is ok because the python lead to more code something like that?

Reply:
The slugify is always done in python and always come with the unique path check, slugifying in js will always be incomplete and there is no way to compute the final url, as it needs to reach all other pages to ensure it is not already taken (and if so act accordingly, something like adding a suffix, I don't have code in front of me right now)

> Add more accuracy when comparing user input to the current page url
to prevent some situations where the redirect is useless:
Example: old page url value = '/home-page' and user value = 'home page'
=> this will lead to the same url = '/home-page' (after slugify on
backend) so the redirect here doesn't make any sense and the redirect
option should not be visible.

I am not sure why would someone rename the current page url "/contact-us" to "contact us".

Why would we want code for handling that? That's overkill optimization, this should be like 1 case out of 10.000 url modification, on top of being weird to do. In that case we shouldn't even care about the redirect (which is not even correct in js anyway as I explained).

Just a quick reply without being able to check the code or blame anything, based on the commit message. So it might have been like that before but I doubt it.

Reply:
@xO-Tx Would be good to have a list of what happened before:
- If I type a URL which already exist, what happens? (does it show the redirect, does it tell you anything, etc)
- If I type a URL which is the same after the slugification (what you tried to do here)
- etc

My opinion without much thinking is that it's probably overkill to try to check if the URL already exist or not (I mean if it is changed automatically after save if it does exist, it seems ok). As for the slugification part `"/contact-us" to "contact us"` I am not sure it is as overkill as you think it is @rdeodoo, I can very well imagine people trying to replace the "-" inside the URL that was automatically added by spaces ^^ But I agree the implementation at least seems overkill (although probably correct even though incomplete as you said).

Anyways: I would really like a list of features that were in 15.0 and how it worked @xO-Tx :+1:

Reply:
Well I didn't check if possible (is the reload occuring after a RPC based on that rpc result?) but maybe the python could simply slug and check if the slug is the same as the httprequest.request.path -> The slug ends up the same as the incoming url, don't reload, or even the JS could do that (if the reload url given by the server is the same as the current page url).

Reply:
@xO-Tx Good explanation, if what you found out on the python side is true I guess it could be worth a fix over there. For the JS side, maybe it's a bit overkill (mainly because of the need to make an RPC etc and the bunch of lines needed to achieve it) but I am not strongly against like @rdeodoo seemed to be, let's let him decide, he knows this flow and code better than me.

Reply:
@rdeodoo 's MESSAGE (MOVED BACK HERE):

> There is no check in backend on URL input, just a JS check:

I am not sure of what we are talking about here TBH. There is python validation of the url input, which is actually checking for the slug and unique path. Maybe we are talking about something else.

About the commit message and what was already discussed, what kind of redirect are we talking about? Browser/js redirect ? Python redirect, or `website.rewrite` redirect?

> It's especially the [A-2] use case that I'm targeting here

So that's about the creation of the `website.rewrite`, right? I tested and indeed there is an useless redirect record created.
That's because the python code was not strong enough and not assuming that this case could exists: if the slugified url of the written url is the same as the already set url, it shouldn't create the redirect record.

So in 15.0, here:
```py
        # Create redirect if needed
        if data['create_redirect']:
            self.env['website.rewrite'].create({
                'name': data['name'],
                'redirect_type': data['redirect_type'],
                'url_from': original_url,
                'url_to': url,
                'website_id': website.id,
            })
``` 
in the `save_page_info()`, it should simply add a check `and original_url != url` with a comment `both url could be the same due to the slug` (better explained than that ofc)
This could be a 14.0 fix.

If that's what we are talking about, it shouldn't be a JS fix but just something better handled in the python/end point side, otherwise the issue and misconfiguration remains possible through other flows

Also, in master, we should introduce an sql constraint about this misconfiguration.

------------------------------------_

Now, reading your **point B**, after having checked and written :point_up: , I guess you were not talking about this issue actually (which should still be cared of I guess).

About that issue, I don't think adding `url = '/' + slugify(url, max_length=1024, path=True)` to check if the slugifies will result in the same url as the current one is ideal, most of the time we will do that twice for no reason (a first time to check that "error" you try to fix, and a second time to get the actual url if we are not in the case you mention).

At best maybe `get_valid_page_url` could take a new param of the pages but that's going too far TBH, unless someone disagrees.

We don't really care if someone rename it's page url from `/contactus` to `/contactus ` (or `/contactus/`) and it becomes `/contactus-1` instead of remaining `/contactus`, that wouldn't actually make more sense, would it? Still not what the user asked.

I think that's a non real use case and we shouldn't introduce code to handle that. It's already a bit complicated as is.

But now that I read the code, I am pretty sure I asked `get_valid_page_url()` to not be added as util but kept duplicate 2 lines of code in a previous PR.
That's probably the way to go here ;-)

```py
def write(self, vals):
    # ...
    if page.url != url:
        url = self.get_valid_page_url(url, website_id)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        copy_param['url'] = self.get_valid_page_url(page_name)

def get_valid_page_url(self, page_url, website_id=False):
    url = '/' + slugify(page_url, max_length=1024, path=True)
    return self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

=>

```py
def write(self, vals):
    # ...
    url = url = '/' + slugify(url, max_length=1024, path=True)
    if page.url != url:
        url =  self.env['website'].with_context(website_id=website_id).get_unique_path(url)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        url = '/' + slugify(page_name, max_length=1024, path=True)
        copy_param['url'] = self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

Reply:
(no I am not talking to myself)

Reply:
> There is no check in backend on URL input, just a JS check:

I am not sure of what we are talking about here TBH. There is python validation of the url input, which is actually checking for the slug and unique path. Maybe we are talking about something else.

About the commit message and what was already discussed, what kind of redirect are we talking about? Browser/js redirect ? Python redirect, or `website.rewrite` redirect?

> It's especially the [A-2] use case that I'm targeting here

So that's about the creation of the `website.rewrite`, right? I tested and indeed there is an useless redirect record created.
That's because the python code was not strong enough and not assuming that this case could exists: if the slugified url of the written url is the same as the already set url, it shouldn't create the redirect record.

So in 15.0, here:
```py
        # Create redirect if needed
        if data['create_redirect']:
            self.env['website.rewrite'].create({
                'name': data['name'],
                'redirect_type': data['redirect_type'],
                'url_from': original_url,
                'url_to': url,
                'website_id': website.id,
            })
``` 
in the `save_page_info()`, it should simply add a check `and original_url != url` with a comment `both url could be the same due to the slug` (better explained than that ofc)
This could be a 14.0 fix.

If that's what we are talking about, it shouldn't be a JS fix but just something better handled in the python/end point side, otherwise the issue and misconfiguration remains possible through other flows

Also, in master, we should introduce an sql constraint about this misconfiguration.

------------------------------------_

Now, reading your **point B**, after having checked and written :point_up: , I guess you were not talking about this issue actually (which should still be cared of I guess).

About that issue, I don't think adding `url = '/' + slugify(url, max_length=1024, path=True)` to check if the slugifies will result in the same url as the current one is ideal, most of the time we will do that twice for no reason (a first time to check that "error" you try to fix, and a second time to get the actual url if we are not in the case you mention).

At best maybe `get_valid_page_url` could take a new param of the pages but that's going too far TBH, unless someone disagrees.

We don't really care if someone rename it's page url from `/contactus` to `/contactus ` (or `/contactus/`) and it becomes `/contactus-1` instead of remaining `/contactus`, that wouldn't actually make more sense, would it? Still not what the user asked.

I think that's a non real use case and we shouldn't introduce code to handle that. It's already a bit complicated as is.

But now that I read the code, I am pretty sure I asked `get_valid_page_url()` to not be added as util but kept duplicate 2 lines of code in a previous PR.
That's probably the way to go here ;-)

```py
def write(self, vals):
    # ...
    if page.url != url:
        url = self.get_valid_page_url(url, website_id)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        copy_param['url'] = self.get_valid_page_url(page_name)

def get_valid_page_url(self, page_url, website_id=False):
    url = '/' + slugify(page_url, max_length=1024, path=True)
    return self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

=>

```py
def write(self, vals):
    # ...
    url = url = '/' + slugify(url, max_length=1024, path=True)
    if page.url != url:
        url =  self.env['website'].with_context(website_id=website_id).get_unique_path(url)
        page.menu_ids.write({'url': url})
    # ...

def clone_page(self, page_id, page_name=None, clone_menu=True):
    # ...
    if page_name:
        url = '/' + slugify(page_name, max_length=1024, path=True)
        copy_param['url'] = self.env['website'].with_context(website_id=website_id).get_unique_path(url)
```

Reply:
@qsm-odoo I deleted the last one and repasted it, will make stuff a (little) bit clearer :joy: 

@xO-Tx Try to replay on discussion on.. discussion, here you started to replay to everyone in a message in the tread, instead of continuing here, it was impossible to understand anything anymore. Thanks @qsm-odoo for moving stuff here.

Reply:
> in the `save_page_info()`, it should simply add a check `and original_url != url` with a comment `both url could be the same due to the slug` (better explained than that ofc)
> This could be a 14.0 fix.
> 
> If that's what we are talking about, it shouldn't be a JS fix but just something better handled in the python/end point side, otherwise the issue and misconfiguration remains possible through other flows

@rdeodoo actually that's the useless `website.rewrite` I was talking about and I added this "JS slugify version" mainly because I saw it was confusing to allow it in first place by making the **"page properties redirect option"** visible while it shouldn't be.

> We don't really care if someone rename it's page url from `/contactus` to `/contactus ` (or `/contactus/`) and it becomes `/contactus-1` instead of remaining `/contactus`, that wouldn't actually make more sense, would it? Still not what the user asked.

Mmm... I thought it was a consistency issue... because this changes everything now:

**C1-** We want `/contactus/` or ... to be `/contactus`:
- We should prevent backend from setting `/contactus-1`
- We should prevent backend from creating a `website.rewrite` to the same URL.
- We should disable the "redirect" option in page properties.

**C2-** It's OK to have `/contactus/` or ... => `/contactus-1`
- It's a different URL, so we can create a `website.rewrite` for it (no problem here).
- Nothing to do on JS.

My goal here was to have the configuration in **C1** but if **C2** is OK for you I'm OK too :+1: 

> But now that I read the code, I am pretty sure I asked get_valid_page_url() to not be added as util but kept duplicate 2 lines of code in a previous PR.

I remember that I tried many times to remove it, I think the last time was a part of the [wizard PR](https://github.com/odoo/odoo/pull/95949/commits/275c04bbb25e9d5900aed651304ea5986aa59507#diff-77bd6b19c39e211959885024bcad914655ff84cfc10c16633687d014e50aa69aL276), but I don't think I removed in other PR after we decided to close it :grimacing:

> ```py
> def write(self, vals):
>     # ...
>     url = url = '/' + slugify(url, max_length=1024, path=True)
>     if page.url != url:
>         url =  self.env['website'].with_context(website_id=website_id).get_unique_path(url)
>         page.menu_ids.write({'url': url})
>     # ...
> 
> def clone_page(self, page_id, page_name=None, clone_menu=True):
>     # ...
>     if page_name:
>         url = '/' + slugify(page_name, max_length=1024, path=True)
>         copy_param['url'] = self.env['website'].with_context(website_id=website_id).get_unique_path(url)
> ```

Indeed, it's better this way :+1:

Reply:
@qsm-odoo I added a small tweak on the code:
- Remove the **"JS slugify code"** and keep only the **"/ in non edited part of the field"**
- Remove the `get_valid_page_url()` (as @rdeodoo suggested) & set focus on the first field in page properties :+1: